cura/CuraPackageManager.py

```diff
@@ -0,0 +1,357 @@
+# Copyright (c) 2018 Ultimaker B.V.
+# Cura is released under the terms of the LGPLv3 or higher.
+
+from typing import Optional
+import json
+import os
+import shutil
+import zipfile
+import tempfile
+
+from PyQt5.QtCore import pyqtSlot, QObject, pyqtSignal
+
+from UM.Logger import Logger
+from UM.Resources import Resources
+
+from cura.Utils import VersionTools
+
+
+class CuraPackageManager(QObject):
+
+    # The prefix that's added to all files for an installed package to avoid naming conflicts with user created
+    # files.
+    PREFIX_PLACE_HOLDER = "-CP;"
+
+    def __init__(self, parent = None):
+        super().__init__(parent)
+
+        self._application = parent
+        self._container_registry = self._application.getContainerRegistry()
+        self._plugin_registry = self._application.getPluginRegistry()
+
+        # JSON file that keeps track of all installed packages.
+        self._package_management_file_path = os.path.join(os.path.abspath(Resources.getDataStoragePath()),
+                                                          "packages.json")
+        self._installed_package_dict = {}  # a dict of all installed packages
+        self._to_remove_package_set = set()  # a set of packages that need to be removed at the next start
+        self._to_install_package_dict = {}  # a dict of packages that need to be installed at the next start
+
+    installedPackagesChanged = pyqtSignal()  # Emitted whenever the installed packages collection have been changed.
+
+    def initialize(self):
+        self._loadManagementData()
+        self._removeAllScheduledPackages()
+        self._installAllScheduledPackages()
+
+    # (for initialize) Loads the package management file if exists
+    def _loadManagementData(self) -> None:
+        if not os.path.exists(self._package_management_file_path):
+            Logger.log("i", "Package management file %s doesn't exist, do nothing", self._package_management_file_path)
+            return
+
+        with open(self._package_management_file_path, "r", encoding = "utf-8") as f:
+            management_dict = json.loads(f.read(), encoding = "utf-8")
+
+            self._installed_package_dict = management_dict["installed"]
+            self._to_remove_package_set = set(management_dict["to_remove"])
+            self._to_install_package_dict = management_dict["to_install"]
+
+            Logger.log("i", "Package management file %s is loaded", self._package_management_file_path)
+
+    def _saveManagementData(self) -> None:
+        with open(self._package_management_file_path, "w", encoding = "utf-8") as f:
```

Comment:
You need to check for the configuration file lock before writing this file, too. Otherwise you will get \x00 characters due to corrupt files when people are running multiple instances of Cura at the same time.

Reply:
@LipuFei

Reply:
Used lock in 1fb2edace